File: backend/src/schemas/collection.py
"""
催缴管理相关的 Pydantic schemas
"""

from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class CollectionMethodEnum(str, Enum):
    """催缴方式枚举"""

    PHONE = "phone"
    SMS = "sms"
    EMAIL = "email"
    WECOM = "wecom"
    VISIT = "visit"
    LETTER = "letter"
    OTHER = "other"


class CollectionStatusEnum(str, Enum):
    """催缴状态枚举"""

    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    SUCCESS = "success"
    FAILED = "failed"
    PARTIAL = "partial"


class CollectionRecordBase(BaseModel):
    """催缴记录基础Schema"""

    ledger_id: str = Field(..., description="关联的租金台账ID")
    contract_id: str = Field(..., description="关联合同ID")
    collection_method: CollectionMethodEnum = Field(..., description="催缴方式")
    collection_date: date = Field(..., description="催缴日期")
    collection_status: CollectionStatusEnum = Field(
        CollectionStatusEnum.PENDING, description="催缴状态"
    )
    contacted_person: str | None = Field(None, description="被联系人")
    contact_phone: str | None = Field(None, description="联系电话")
    promised_amount: Decimal | None = Field(None, ge=0, description="承诺付款金额")
    promised_date: date | None = Field(None, description="承诺付款日期")
    actual_payment_amount: Decimal | None = Field(
        None, ge=0, description="实际付款金额"
    )
    collection_notes: str | None = Field(None, description="催缴备注")
    next_follow_up_date: date | None = Field(None, description="下次跟进日期")


class CollectionRecordCreate(CollectionRecordBase):
    """创建催缴记录Schema"""

    operator: str | None = Field(None, description="操作人")
    operator_id: str | None = Field(None, description="操作人ID")


class CollectionRecordUpdate(BaseModel):
    """更新催缴记录Schema"""

    collection_status: CollectionStatusEnum | None = Field(None, description="催缴状态")
    contacted_person: str | None = Field(None, description="被联系人")
    contact_phone: str | None = Field(None, description="联系电话")
    promised_amount: Decimal | None = Field(None, ge=0, description="承诺付款金额")
    promised_date: date | None = Field(None, description="承诺付款日期")
    actual_payment_amount: Decimal | None = Field(
        None, ge=0, description="实际付款金额"
    )
    collection_notes: str | None = Field(None, description="催缴备注")
    next_follow_up_date: date | None = Field(None, description="下次跟进日期")


class CollectionRecordResponse(CollectionRecordBase):
    """催缴记录响应Schema"""

    id: str
    operator: str | None = Field(None, description="操作人")
    operator_id: str | None = Field(None, description="操作人ID")
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CollectionTaskSummary(BaseModel):
    """催缴任务汇总"""

    total_overdue_count: int = Field(..., description="逾期台账总数")
    total_overdue_amount: Decimal = Field(..., description="逾期总金额")
    pending_collection_count: int = Field(..., description="待催缴数量")
    this_month_collection_count: int = Field(..., description="本月催缴次数")
    collection_success_rate: Decimal | None = Field(None, description="催缴成功率")


# 分页Schema
class CollectionRecordListResponse(BaseModel):
    """催缴记录列表响应"""

    items: list[CollectionRecordResponse]
    total: int
    page: int
    limit: int
    pages: int

File: frontend/src/types/collection.ts
/**
 * 催缴管理相关类型定义
 * 对齐: backend/src/schemas/collection.py
 */

/** 催缴方式枚举 */
export enum CollectionMethod {
  PHONE = 'phone',
  SMS = 'sms',
  EMAIL = 'email',
  WECOM = 'wecom',
  VISIT = 'visit',
  LETTER = 'letter',
  OTHER = 'other',
}

/** 催缴方式显示名称 */
export const CollectionMethodLabels: Record<CollectionMethod, string> = {
  [CollectionMethod.PHONE]: '电话',
  [CollectionMethod.SMS]: '短信',
  [CollectionMethod.EMAIL]: '邮件',
  [CollectionMethod.WECOM]: '企业微信',
  [CollectionMethod.VISIT]: '上门拜访',
  [CollectionMethod.LETTER]: '信函',
  [CollectionMethod.OTHER]: '其他',
};

/** 催缴状态枚举 */
export enum CollectionStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  SUCCESS = 'success',
  FAILED = 'failed',
  PARTIAL = 'partial',
}

/** 催缴状态显示名称 */
export const CollectionStatusLabels: Record<CollectionStatus, string> = {
  [CollectionStatus.PENDING]: '待催缴',
  [CollectionStatus.IN_PROGRESS]: '催缴中',
  [CollectionStatus.SUCCESS]: '催缴成功',
  [CollectionStatus.FAILED]: '催缴失败',
  [CollectionStatus.PARTIAL]: '部分成功',
};

/** 催缴状态颜色 */
export const CollectionStatusColors: Record<CollectionStatus, string> = {
  [CollectionStatus.PENDING]: 'default',
  [CollectionStatus.IN_PROGRESS]: 'processing',
  [CollectionStatus.SUCCESS]: 'success',
  [CollectionStatus.FAILED]: 'error',
  [CollectionStatus.PARTIAL]: 'warning',
};

/** 催缴记录基础类型 */
export interface CollectionRecordBase {
  ledger_id: string;
  contract_id: string;
  collection_method: CollectionMethod;
  collection_date: string; // ISO date string
  collection_status: CollectionStatus;
  contacted_person?: string | null;
  contact_phone?: string | null;
  promised_amount?: number | null;
  promised_date?: string | null;
  actual_payment_amount?: number | null;
  collection_notes?: string | null;
  next_follow_up_date?: string | null;
}

/** 创建催缴记录请求 */
export interface CollectionRecordCreate extends CollectionRecordBase {
  operator?: string | null;
  operator_id?: string | null;
}

/** 更新催缴记录请求 */
export interface CollectionRecordUpdate {
  collection_status?: CollectionStatus | null;
  contacted_person?: string | null;
  contact_phone?: string | null;
  promised_amount?: number | null;
  promised_date?: string | null;
  actual_payment_amount?: number | null;
  collection_notes?: string | null;
  next_follow_up_date?: string | null;
}

/** 催缴记录响应 */
export interface CollectionRecord extends CollectionRecordBase {
  id: string;
  operator?: string | null;
  operator_id?: string | null;
  created_at: string;
  updated_at: string;
}

/** 催缴任务汇总 */
export interface CollectionTaskSummary {
  total_overdue_count: number;
  total_overdue_amount: number;
  pending_collection_count: number;
  this_month_collection_count: number;
  collection_success_rate?: number | null;
}

/** 催缴记录列表响应 */
export interface CollectionRecordListResponse {
  items: CollectionRecord[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}
